bindgen/ir/function.rs

```diff
@@ -674,25 +683,35 @@ impl ClangSubItemParser for Function {
             return Err(ParseError::Continue);
         }
 
+        let linkage = cursor.linkage();
+        let linkage = match linkage {
+            CXLinkage_External | CXLinkage_UniqueExternal => Linkage::External,
+            CXLinkage_Internal => Linkage::Internal,
+            _ => return Err(ParseError::Continue),
+        };
+
         if cursor.is_inlined_function() ||
             cursor
                 .definition()
                 .map_or(false, |x| x.is_inlined_function())
         {
-            if !context.options().generate_inline_functions {
+            if !(context.options().generate_inline_functions ||
+                context.options().wrap_non_extern_fns)
+            {
                 return Err(ParseError::Continue);
             }
+
             if cursor.is_deleted_function() {
                 return Err(ParseError::Continue);
             }
-        }
 
-        let linkage = cursor.linkage();
-        let linkage = match linkage {
-            CXLinkage_External | CXLinkage_UniqueExternal => Linkage::External,
-            CXLinkage_Internal => Linkage::Internal,
-            _ => return Err(ParseError::Continue),
-        };
+            if context.options().wrap_non_extern_fns &&
+                cursor.is_inlined_function() &&
+                matches!(linkage, Linkage::External)
+            {
```

Comment:
I'm a bit confused about this condition, when is an inline function extern (can it even be?), and why do we need to bail here?

A comment would be useful.

Reply:
yeah I don't understand very well what's happening but what I've seen is this. If you have this header:
```c
static inline int foo() {
    return 0;
}
static int bar() {
    return 1;
}
inline int baz() {
    return 2;
}
```
`baz` is a extern (non-static) inlined function. If you generate the following `c` wrappers:
```c
extern "C" {
    #[link_name = "\u{1}foo__extern"]
    pub fn foo() -> ::std::os::raw::c_int;
}
extern "C" {
    #[link_name = "\u{1}bar__extern"]
    pub fn bar() -> ::std::os::raw::c_int;
}
extern "C" {
    #[link_name = "\u{1}baz__extern"]
    pub fn baz() -> ::std::os::raw::c_int;
}
```
and compile the wrappers including the original headers without optimizations using `clang` you get the following object file:
```objdump
Disassembly of section .text:

0000000000000000 <foo__extern>:
   0:	55                   	push   %rbp
   1:	48 89 e5             	mov    %rsp,%rbp
   4:	e8 27 00 00 00       	call   30 <foo>
   9:	5d                   	pop    %rbp
   a:	c3                   	ret
   b:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

0000000000000010 <bar__extern>:
  10:	55                   	push   %rbp
  11:	48 89 e5             	mov    %rsp,%rbp
  14:	e8 27 00 00 00       	call   40 <bar>
  19:	5d                   	pop    %rbp
  1a:	c3                   	ret
  1b:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

0000000000000020 <baz__extern>:
  20:	55                   	push   %rbp
  21:	48 89 e5             	mov    %rsp,%rbp
  24:	e8 00 00 00 00       	call   29 <baz__extern+0x9>
  29:	5d                   	pop    %rbp
  2a:	c3                   	ret
  2b:	0f 1f 44 00 00       	nopl   0x0(%rax,%rax,1)

0000000000000030 <foo>:
  30:	55                   	push   %rbp
  31:	48 89 e5             	mov    %rsp,%rbp
  34:	31 c0                	xor    %eax,%eax
  36:	5d                   	pop    %rbp
  37:	c3                   	ret
  38:	0f 1f 84 00 00 00 00 	nopl   0x0(%rax,%rax,1)
  3f:	00 

0000000000000040 <bar>:
  40:	55                   	push   %rbp
  41:	48 89 e5             	mov    %rsp,%rbp
  44:	b8 01 00 00 00       	mov    $0x1,%eax
  49:	5d                   	pop    %rbp
  4a:	c3                   	ret
```
So `baz__extern` is trying to call `baz` but `baz` is nowhere to be found. However, if you pass `-O2` to clang, you get this:
```objdump
Disassembly of section .text:

0000000000000000 <foo__extern>:
   0:	31 c0                	xor    %eax,%eax
   2:	c3                   	ret
   3:	66 66 66 66 2e 0f 1f 	data16 data16 data16 cs nopw 0x0(%rax,%rax,1)
   a:	84 00 00 00 00 00 

0000000000000010 <bar__extern>:
  10:	b8 01 00 00 00       	mov    $0x1,%eax
  15:	c3                   	ret
  16:	66 2e 0f 1f 84 00 00 	cs nopw 0x0(%rax,%rax,1)
  1d:	00 00 00 

0000000000000020 <baz__extern>:
  20:	b8 02 00 00 00       	mov    $0x2,%eax
  25:	c3                   	ret
```

where `baz` is correctly inlined and works as expected.

So to be on the safe side I decided to omit non-static inlined functions.

I wouldn't be sure how to explain this behavior other than saying that we only generate wrappers for `static` and `static inline` functions but not `inline` functions due to possible linking problems.

Reply:
OK apparently `inline` functions that are not `static` require an special `extern` declaration so the compiler can actually find them. So I guess the "right" name for this feature is `wrap-static-fns` as we can only handle `static` and `static inline` functions.